watertap/core/tests/test_wt_database.py

```diff
@@ -0,0 +1,156 @@
+                           match="Could not find entry for foo in database."):
+            db._get_technology("foo")
+
+        assert len(db._cached_files) == 1
+        assert "foo" not in db._cached_files
+
+    @pytest.mark.unit
+    def test_get_unit_operation_parameters_default(self, db):
+        data = db.get_unit_operation_parameters("nanofiltration")
+
+        assert data == db._cached_files["nanofiltration"]["default"]
+
+        # Check for a few expected keys to check what we got back looks right
+        assert "constituent_removals" in data
+        assert "electricity_intensity" in data
+        assert "recovery" in data
+
+    @pytest.mark.unit
+    def test_get_unit_operation_parameters_invalid_subtype(self, db):
+        with pytest.raises(KeyError,
+                           match="Received unrecognised subtype foo for "
+                           "technology nanofiltration."):
+            db.get_unit_operation_parameters("nanofiltration", subtype="foo")
+
+    @pytest.mark.unit
+    def test_get_unit_operation_parameters_single_subtype(self, db):
+        # First, insert some data for a subtype into nanofiltration entry
+        db._cached_files["nanofiltration"]["subtype1"] = {
+            "recovery": "overloaded",
+            "new_param": True}
+
+        # Load data for subtype
+        data = db.get_unit_operation_parameters(
+            "nanofiltration", subtype="subtype1")
+
+        # Check data
+        for k, v in data.items():
+            if k == "recovery":
+                assert v == "overloaded"
+            elif k == "new_param":
+                assert v is True
+            else:
+                # All other entries should match defaults
+                assert v == db._cached_files["nanofiltration"]["default"][k]
+
+    @pytest.mark.unit
+    def test_get_unit_operation_parameters_invalid_list_of_subtype(self, db):
+        with pytest.raises(KeyError,
+                           match="Received unrecognised subtype foo for "
+                           "technology nanofiltration."):
+            db.get_unit_operation_parameters(
+                "nanofiltration", subtype=["subtype1", "foo"])
+
+    @pytest.mark.unit
+    def test_get_unit_operation_parameters_multi_subtype(self, db):
+        # First, insert some data for a 2nd subtype into nanofiltration entry
+        db._cached_files["nanofiltration"]["subtype2"] = {
+            "recovery": "overloaded_again",
+            "new_param_2": False}
+
+        # Load data for subtype
+        data = db.get_unit_operation_parameters(
+            "nanofiltration", subtype="subtype2")
+
+        # Check data
+        for k, v in data.items():
+            if k == "recovery":
+                assert v == "overloaded_again"
+            elif k == "new_param":
+                assert v is True
+            elif k == "new_param_2":
+                assert v is False
+            else:
+                # All other entries should match defaults
```

Comment:
The test is fine, but do we want to keep parameters from default that aren't necessarily included in said subtype? In this example, what if `new_param` is not a relevant parameter in `subtype2`, for example, removal of a constituent which is not applicable to the subtype? 

I suppose one way to handle this would be to assign a NaN or something like that in the yaml file which would overload `new_param` under `subtype2`.

Reply:
Note that the unit model still gets to decide which parameters to use, so it doesn't matter if we return more parameters than it needs.

Reply:
In that case, I guess one would just need to be careful when loading those parameters to a unit model.

Reply:
I did this so that a subtype only needs to contain values for parameters that change, and to have a fall back in case something isn't there.